erp-playbook/library/mr_provisioner_power.py

```diff
@@ -0,0 +1,205 @@
+    url = urljoin(url, "/api/v1/machine?q={}&show_all=false".format(q))
+    r = requests.get(url, headers=headers)
+    if r.status_code != 200:
+        raise ProvisionerError('Error fetching {}, HTTP {} {}'.format(url, r.status_code, r.reason))
+    if len(r.json()) == 0:
+        raise ProvisionerError('Error no assigned machine found with name "{}"'.format(machine_name))
+    if len(r.json()) > 1:
+        raise ProvisionerError('Error more than one machine found with name "{}", {}'.format(machine_name, r.json()))
+    return r.json()[0]
+
+
+def get_power_state(url, token, machine_id):
+    """ Look up power state by machine ID """
+    headers = {'Authorization': token}
+    url = urljoin(url, "/api/v1/machine/{}/power".format(machine_id))
+    r = requests.get(url, headers=headers)
+    if r.status_code != 200:
+        raise ProvisionerError('Error posting {}, HTTP {} {}'.format(url, r.status_code, r.reason))
+    return r.json()
+
+
+def change_power_state(url, token, machine_id, state):
+    """ Change power state by machine ID """
+    headers = {'Authorization': token}
+    url = urljoin(url, "/api/v1/machine/{}/power".format(machine_id))
+    data = json.dumps({'state': state})
+    r = requests.post(url, headers=headers, data=data)
+    if r.status_code not in [200, 202]:
+        raise ProvisionerError('Error posting {}, HTTP {} {}'.format(url, r.status_code, r.reason))
+
+
+def run_module():
+    module_args = dict(
+        machine_name=dict(type='str', required=True),
+        url=dict(type='str', required=True),
+        token=dict(type='str', required=True),
+        wait_for_power=dict(type='str', choices=['on', 'off'], required=True),
+        timeout=dict(type='int', default=21600),
+        sleep=dict(type='int', default=300),
+        force_power=dict(type='str', choices=['on', 'off']),
+    )
+
+    result = dict(
+        changed=False,
+        debug={},
+    )
+
+    module = AnsibleModule(
+        argument_spec=module_args,
+        supports_check_mode=True
+    )
+
+    if module.check_mode:
+        return result
+
+    # Look up machine, verify assignment
+    try:
+        machine = get_machine_by_name(module.params['url'],
+                                      module.params['token'],
+                                      module.params['machine_name'])
+    except ProvisionerError as e:
+        module.fail_json(msg=str(e), **result)
+
+    # Wait for power on/off state with timeout.
+    max_end_time = datetime.utcnow() + timedelta(seconds=module.params['timeout'])
+    while datetime.utcnow() < max_end_time:
+        try:
+            power_state = get_power_state(module.params['url'],
+                                          module.params['token'],
+                                          machine_id=machine['id'])
+
+        except ProvisionerError as e:
+            module.fail_json(msg=str(e), **result)
+
+        if power_state['state'] == module.params['force_power']:
```

Comment:
shouldn't this check wait_for_power instead of force_power?

Reply:
Fixed.